Bug/697 different EUI formats

bastbu opened the pull request:
# PR for issue #697

## What is being addressed

With this PR we introduce different formats for all the `Eui`.

- `G`, `D`, `null` format give us dashed Hex format (88-99-AA-BB-CC-DD-EE-FF)
- `N` gives us non-dashed (similar as for `Guid`)
- `I` gives us ID6 (8899:AABB:CCDD:EEFF)
- `E` gives us the ID6 format (88:99:AA:BB:CC:DD:EE:FF)
- `e`, `n`, `i`, `d` give their respective formats using lowercase alpha digits

## How is this addressed

We add a `TryParse(string format)`, `Parse(string format)` and `ToString(string format, IFormatProvider formatProvider)` implementation for all euis.

atifaziz reviewed (commented):
This is fantastic work. I would just suggest the following improvements:

- Extract bulk of formatting and parsing into non-generated code (see 5e3888274adb1139202df80aed47eab1f4d6e247 for an example of formatting).
- Remove the need for a format specification during parsing and just auto-detect the format to parse assuming there is no ambiguity (think there isn't).
- Add tests to exercise integration with either `string.Format` or string interpolation (`$"..."`).

bastbu commented:
Thanks for the review. I addressed the first two points with d09f2528c66b1b44dd35286622483e73a6d6d91b and added the test for string interpolation in 225977d2a4de386f7e0c7c080b209605c577bbef

atifaziz commented on Tests/Unit/LoRaWanTests/EuiTests.cs:
```diff
@@ -127,10 +174,22 @@ public void TryParse_Input_Is_Case_Insensitive()
+            Assert.Equal(default, result);
         }
+
+        [Fact]
+        public void String_Interpolation_Success_Case()
+        {
+            var result = $"N = {Subject:N}, I = {Subject:I}";
+            Assert.Equal("N = 0123456789ABCDEF, I = 0123:4567:89AB:CDEF", result);
+        }
```
I would recommend testing all formats here in the interpolation.

Alternatively, you could turn this into a theory, borrowing almost the same data as you have for `ToString_Returns_Correctly_Formatted_String`:

```suggestion
        [Theory]
        [InlineData("{0}", "01-23-45-67-89-AB-CD-EF")]
        [InlineData("{0:G}", "01-23-45-67-89-AB-CD-EF")]
        [InlineData("{0:D}", "01-23-45-67-89-AB-CD-EF")]
        [InlineData("{0:g}", "01-23-45-67-89-ab-cd-ef")]
        [InlineData("{0:d}", "01-23-45-67-89-ab-cd-ef")]
        [InlineData("{0:E}", "01:23:45:67:89:AB:CD:EF")]
        [InlineData("{0:e}", "01:23:45:67:89:ab:cd:ef")]
        [InlineData("{0:I}", "0123:4567:89AB:CDEF")]
        [InlineData("{0:i}", "0123:4567:89ab:cdef")]
        [InlineData("{0:N}", "0123456789ABCDEF")]
        [InlineData("{0:n}", "0123456789abcdef")]
        public void String_Interpolation_Success_Case(string format, string expected)
        {
            var result = string.Format(CultureInfo.InvariantCulture, format, Subject);
            Assert.Equal(expected, result);
        }
```

atifaziz reviewed (approved):
LGTM

bastbu commented on LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/Eui.g.tt:
```diff
@@ -36,14 +36,7 @@ namespace LoRaWan
-            Hexadecimal.Write(bytes, chars, '-');
-            return new string(chars);
-        }
+        public override string ToString() => ToString("G", null);
```
Solved in 5f883a5128e8945ada9b07d494345354b5601f31

bastbu commented on Tests/Unit/LoRaWanTests/EuiTests.cs:
```diff
@@ -19,6 +20,7 @@ public abstract class EuiTests<T> where T : struct, IEquatable<T>
 
         private static readonly Func<T, T, bool> Equal = Operators<T>.Equality;
         private static readonly Func<T, T, bool> NotEqual = Operators<T>.Inequality;
+        private static readonly char?[] SupportedFormats = new char?[] { null, 'G', 'g', 'D', 'd', 'I', 'i', 'N', 'n', 'E', 'e' };
```
Solved in 4eeb29e575c2becd76e82a6eee90bc81853fe8c3

bastbu replied:
and 6daf28c32ee329fcb352afc5b57583f4409f458b
